pkg/inclusion/paths.go

```diff
@@ -0,0 +1,123 @@
+package inclusion
+
+// genSubTreeRootPath calculates the path to a given subtree root of a node, given the
+// depth and position of the node. note: the root of the tree is depth 0.
+// The following nolint can be removed after this function is used.
+//nolint:unused,deadcode
+func genSubTreeRootPath(depth int, pos uint) []bool {
+	path := make([]bool, depth)
+	for i := depth; i >= 0; i-- {
+		if (pos & (1 << i)) == 0 {
+			path = append(path, false)
+		} else {
+			path = append(path, true)
+		}
+	}
+	return path
+}
+
+// coord identifies a tree node using the depth and position
+// Depth       Position
+// 0              0
+//               / \
+//              /   \
+// 1           0     1
+//            /\     /\
+// 2         0  1   2  3
+//          /\  /\ /\  /\
+// 3       0 1 2 3 4 5 6 7
+type coord struct {
+	// depth is the typical depth of a tree, 0 being the root
+	depth uint64
+	// position is the index of a node of a given depth, 0 being the left most
+	// node
+	position uint64
+}
+
+// climb is a state transition function to simulate climbing a balanced binary
+// tree, using the current node as input and the next highest node as output.
+func (c coord) climb() coord {
+	return coord{
+		depth:    c.depth - 1,
+		position: c.position / 2,
+	}
+}
+
+// canClimbRight uses the current position to calculate the direction of the next
+// climb. Returns true if the next climb is right (if the position (index) is
+// even).
+func (c coord) canClimbRight() bool {
+	return c.position%2 == 0
```

Comment:
you're right, we probably should just to avoid potential leaky abstractions (although I hope others do not use coord for anything :sweat_smile:) as we're relying on the implementation to do this for us atm

Reply:
https://github.com/celestiaorg/celestia-app/pull/621/commits/32918de7e0959e5f647c2639ee114adfd15f1217